scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5082 @@
+        Instantiate a distribution with the desired parameters:
+
+        >>> import numpy as np
+        >>> from scipy import stats
+        >>> X = stats.Uniform(a=-0.5, b=0.5)
+
+        Evaluate the inverse log-CDF at the desired argument:
+
+        >>> X.ilogcdf(-0.25)
+        0.2788007830714034
+        >>> np.allclose(X.ilogcdf(-0.25), X.icdf(np.exp(-0.25)))
+        True
+
+        This function may return NaN when the spacing between floating point
+        values is too coarse to represent an argument for which the log-CDF is
+        close to the provided `x`, even when the function is theoretically
+        continuous within the domain of the argument.
+
+        >>> X.ilogcdf(-np.inf), X.ilogcdf(-1e10), X.ilogcdf(-30.)
+        (-0.5, nan, -0.4999999999998863)
+
+        """
+        return self._ilogcdf_dispatch(x, method=method, **self._parameters)
+
+    @_dispatch
+    def _ilogcdf_dispatch(self, x, method=None, **kwargs):
+        if self._overrides('_ilogcdf_formula'):
+            method = self._ilogcdf_formula
+        elif self._overrides('_ilogccdf_formula'):
+            method = self._ilogcdf_complementarity
+        else:
+            method = self._ilogcdf_inversion
+        return method
+
+    def _ilogcdf_formula(self, x, **kwargs):
+        raise NotImplementedError(self._not_implemented)
+
+    def _ilogcdf_complementarity(self, x, **kwargs):
+        return self._ilogccdf_dispatch(_log1mexp(x), **kwargs)
+
+    def _ilogcdf_inversion(self, x, **kwargs):
+        return self._solve_bounded(self._logcdf_dispatch, x, kwargs=kwargs)
+
+    @_set_invalid_nan
+    def icdf(self, x, *, method=None):
+        r"""Inverse cumulative distribution function
+
+        The inverse cumulative distribution function :math:`F^{-1}(x)` is the
+        argument :math:`y` for which the cumulative distribution function
+        :math:`F(y)` evaluates to :math:`x`.
+
+        .. math::
+
+            F^{-1}(x) = y \text{ such that } F(y) = x`
+
+        ``icdf`` accepts `x` for :math:`x \in [0, 1]`.
+
+        Parameters
+        ----------
+        x : Array
+            The argument of the inverse cumulative distribution function
+            (inverse CDF).
+        method : {None, 'formula', 'complementarity', 'inversion'}
+            The strategy used to evaluate the inverse CDF.
+            By default (``None``), the infrastructure chooses between the
+            following options.
+
+            - ``'formula'``: use a formula for the inverse CDF itself
+            - ``'complementarity'``: evaluate the inverse CCDF at the
+                                     complement of `x`; i.e. ``1 - x``.
+            - ``'inversion'``: solve numerically for the argument at which the
+                               CDF is equal to `x`.
+
+            Not all `method` options are available for all distributions.
+            If the selected `method` is not available, a ``NotImplementedError``
+            will be raised.
```

Comment:
Yes, essentially. Note that I didn't intend to document the internal method resolution order here, though. As written, this just specifies that the default picks one, and I happened to list the various methods in order of priority. The reason for not specifying is that I want to maintain flexibility. There have been complaints in the past about the infrastructure not being smart about returning accurate answers when it could, and we have the flexibility to do that here. Regardless of our opinions about whether the infrastructure *should* do smarter things, I want to keep the options open.

Reply:
Let's merge this discussion with https://github.com/mdhaber/scipy/pull/110#discussion_r1551387228. I'll take the libery of resolving this one.